src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,629 @@
+                                              &w->resolve_retry_event_source,
+                                              CLOCK_MONOTONIC,
+                                              /* cap at ~25s */
+                                              now(CLOCK_MONOTONIC) + (2 << MAX(w->retries, (unsigned)7)) * 100 * USEC_PER_MSEC,
+                                              0,
+                                              on_resolve_retry,
+                                              netdev);
+                        if (r < 0)
+                                log_netdev_warning_errno(netdev, r, "Could not arm resolve retry handler: %m");
+                }
+        }
+
+        return 0;
+}
+
+static void resolve_endpoints(NetDev *netdev) {
+        int r, i = 0;
+        Wireguard *w;
+        WireguardEndpoint *endpoint;
+        static const struct addrinfo hints = {
+                .ai_family = AF_UNSPEC,
+                .ai_socktype = SOCK_DGRAM,
+                .ai_protocol = IPPROTO_UDP
+        };
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        LIST_FOREACH(endpoints, endpoint, w->unresolved_endpoints) {
+                if (++i >= MAX_CONCURRENT_QUERIES)
+                        break;
+
+                endpoint->netdev = netdev;
+                r = sd_resolve_getaddrinfo(netdev->manager->resolve,
+                                           &w->resolve_query,
+                                           endpoint->host,
+                                           endpoint->port,
+                                           &hints,
+                                           wireguard_resolve_handler,
+                                           endpoint);
+
+                if (r < 0)
+                        log_netdev_error_errno(netdev, r, "Failed create resolver: %m");
+        }
+}
+
+
+static int netdev_wireguard_post_create(NetDev *netdev, Link *link, sd_netlink_message *m) {
+        Wireguard *w;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        resolve_endpoints(netdev);
+        return 0;
+}
+
+int config_parse_wireguard_listen_port(const char *unit,
+                                       const char *filename,
+                                       unsigned line,
+                                       const char *section,
+                                       unsigned section_line,
+                                       const char *lvalue,
+                                       int ltype,
+                                       const char *rvalue,
+                                       void *data,
+                                       void *userdata) {
+        Wireguard *w;
+        uint16_t port = 0;
+        int r;
+
+        assert(rvalue);
+        assert(data);
+
+        w = WIREGUARD(data);
+        assert(w);
+
+        if (!streq(rvalue, "auto")) {
```

Comment:
that was an request by @poettering and you agreed at that point in time: https://github.com/systemd/systemd/pull/4191#discussion_r87655275